File: Week_02/sliding_window_maximum/Solution.java
package sliding_window_maximum;

import java.util.*;

/**
 * 暴力法
 * 时间复杂度：O(n*k)
 * 空间复杂度：O(n-k+1)
 */
class Solution1 {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int resultSize = nums.length - k + 1;
        int[] result = new int[resultSize];
        for (int i = 0; i < resultSize; i++) {
            int max = Integer.MIN_VALUE;
            for (int j = i; j < i + k; j++)
                max = Math.max(max, nums[j]);
            result[i] = max;
        }
        return result;
    }
}

/**
 * 利用队列，面向过程的写法
 * 时间复杂度：O(n)
 * 空间复杂度：O(n-k+1) + O(k) = O(n)
 */
class Solution2 {

    private class MaxQueue {
        Deque<Integer> queue = new ArrayDeque<>();

        void push(int data) {
            while (!queue.isEmpty() && queue.peekLast() < data)
                queue.pollLast();
            queue.offerLast(data);
        }

        void pop(int data) {
            if (queue.peekFirst() == data) queue.pollFirst();
        }

        int max() {
            return queue.peekFirst();
        }
    }

    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] result = new int[nums.length - k + 1];
        MaxQueue window = new MaxQueue();

        for (int i = 0; i < nums.length; i++) {
            if (i < k - 1) {
                window.push(nums[i]);
            } else {
                window.push(nums[i]);
                result[i - k + 1] = window.max();
                window.pop(nums[i - k + 1]);
            }
        }
        return result;
    }

}

/**
 * 动态规划
 * 时间复杂度：O(n)
 * 空间复杂度：O(n-k+1) + O(n) + O(n) = O(n)
 */
class Solution3 {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;

        int[] maxLeft = new int[n];
        maxLeft[0] = nums[0];

        int[] maxRight = new int[n];
        maxRight[n - 1] = nums[n - 1];

        //Traverse the list from start to end and calculate max_so_far. Reset max after each block boundary (of k elements)
        for (int i = 1; i < n; i++) {
            maxLeft[i] = (i % k == 0) ? nums[i] : Math.max(maxLeft[i - 1], nums[i]);
        }
        //Similarly calculate max in future by traversing from end to start
        for (int i = n - 2; i >= 0; i--) {
            maxRight[i] = ((i + 1) % k == 0) ? nums[i] : Math.max(maxRight[i + 1], nums[i]);
        }

        //sliding max at each position i in current window, sliding-max(i) = max{right_max(i), left_max(i+w-1)}
        int[] result = new int[n - k + 1];
        for (int i = 0; i < n - k + 1; i++) {
            result[i] = Math.max(maxLeft[i + k - 1], maxRight[i]);
        }
        return result;
    }
}

File: Week_02/sliding_window_maximum/Solution.py
import sys
from typing import List


# 暴力法
class Solution1:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        n = len(nums) - k + 1
        for i in range(n):
            max_num = -sys.maxsize  # minus maxsize to get minsize (no minsize in python)
            for j in range(i, i + k):
                max_num = max(max_num, nums[j])
            res.append(max_num)
        return res


# 使用只保留最大值的队列
class Solution2:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:

        class MaxQueue:
            def __init__(self):
                self.q = []

            def push(self, num):
                while self.q and self.q[-1] < num: self.q.pop()
                self.q.append(num)

            def pop(self, num):
                if self.q[0] == num: self.q.pop(0)

            def get_max(self):
                return self.q[0]

        result = []
        window = MaxQueue(nums[:k - 1])
        for i in range(k - 1, len(nums)):
            if i < k - 1:
                window.push(nums[i])
            else:
                window.push(nums[i])
                result.append(window.get_max())
                window.pop(nums[i - k + 1])

        return result


# 动态规划
class Solution2:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]: ...
